Local up cluster enhancements

pohly opened the pull request:
**What this PR does / why we need it**:

Networking in a cluster brought up by hack/local-up-cluster.sh was broken because kube-proxy fails to start: connectivity to the Internet is okay, but anything involving communication among services (for example, starting cluster DNS) fails.

Debugging this took a while because there was no indication in the output of hack/local-up-cluster.sh that something had gone wrong. There were also some harmless "no such process" errors during cleanup.

The kube-proxy startup issue was fixed in parallel by PR #60431 but that fix turned out to be incomplete, so this PR uses a different approach, suppresses the harmless error messages (caused by a race) and also makes it more obvious when processes die by checking them regularly.

**Special notes for your reviewer**:

I used

     KUBE_ENABLE_CLUSTER_DNS=true NET_PLUGIN=kubenet hack/local-up-cluster.sh -O

to start the cluster. Check the kube-dns pod and its logs to see how kube-dns fails to communicate with the API server.

I've left out health checking of kubelet when run under Docker because that isn't something that I have used myself yet and I wanted to send the enhancements without further delays. I can try to add that later once you agree that such a health check is useful.

**Release note**:
```release-note
fix network setup in hack/local-up-cluster.sh (https://github.com/kubernetes/kubernetes/pull/60431)
better error diagnostics in hack/local-up-cluster.sh output
```

pohly committed:
local-up-cluster: avoid "No such process" messages when cleaning up

Killing processes in cleanup is racing with those processes
terminating by themselves when aborting local-up-cluster.sh with
CTRL-C. That leads to unnecessary error messages:

  ^CCleaning up...
  kill: (21592): No such process
  kill: (21586): No such process

pohly committed:
local-up-cluster: warn about failing processes

This helps developers detect when something went wrong and then also
points them in the right direction by showing where the corresponding
log file is, if there is one.

For example, since commit 43cb024402e6 in April 2017 kube-proxy failed
to start because featureGates was not set correctly in the .yaml file,
but no-one seems to have noticed this.

pohly committed:
local-up-cluster: fix kube-proxy featureGates configuration

Commit 43cb024402e6 replaced command line parameters with a .yaml
configuration file. But feature gates must be configured with a map in
.yaml, not with a comma-separated string as in the command line
parameters. As a result, kube-proxy failed to start and networking was
broken in the cluster.

Commit c339fc0c4fad tried to fix that by moving feature gates back to
the command line, but later it was found out that the command line
parameter gets ignored when also specifying a config.

Therefore now the feature gates variable gets converted into a proper
map in the config.

pohly commented:
/cc @ncdc

dims commented:
@pohly the kube-proxy problem was fixed already https://github.com/kubernetes/kubernetes/pull/60431 i believe. yes, the other 2 commits in this PR are valuable for sure!

pohly commented:
On Thu, 2018-03-01 at 01:42 -0800, Davanum Srinivas wrote:
> @pohly the kube-proxy problem was fixed already #60431 i believe. 

Right. I had rebased onto master before submitting, but because that
went cleanly missed that someone had edited the file in the meantime.
What are the chances that this is broken for so long and then two
people debug and fix it concurrently? ;-}

#60431 looks like a cleaner fix, so I'll just remove my change.

> yes, the other 2 commits in this PR are valuable for sure!

Let's keep them then in this PR.

Force-pushed and updated PR description (in particular the release
notes).

dims commented:
/ok-to-test

dims commented:
@pohly Please see https://github.com/kubernetes/kubernetes/pull/60431#issuecomment-369624579 apparently i was wrong in my patch. So please resurrect yours :) (Thanks @ncdc)

pohly commented:
On Thu, 2018-03-01 at 15:54 +0000, Davanum Srinivas wrote:
> @pohly Please see #60431 (comment) apparently i was wrong in my
> patch. So please resurrect yours :) (Thanks @ncdc)

"git reflog" to the rescue... ;-)

I've re-added a slightly updated commit. Please have another look.

kube-proxy seems to pick up the feature gates correctly:

$ FEATURE_GATES="MountPropagation=true,CSIPersistentVolume=true" hack/local-up-cluster.sh -O
...

$ grep CSIPersistent /tmp/kube-proxy.log 
I0301 17:10:49.026534   44807 feature_gate.go:226] feature gates: &{map[CSIPersistentVolume:true MountPropagation:true]}
$ cat /tmp/kube-proxy.yaml 
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
clientConnection:
  kubeconfig: /var/run/kubernetes/kube-proxy.kubeconfig
hostnameOverride: 127.0.0.1
mode: 
featureGates:
  MountPropagation: true
  CSIPersistentVolume: true

dims commented:
LGTM 👍

dims commented:
/assign @ixdy 
/assign @cblecker

ixdy commented on hack/local-up-cluster.sh:
```diff
@@ -400,6 +400,44 @@ cleanup()
 
+# Check if all processes are still running. Prints a warning once each time
+# a process dies unexpectedly.
+healthcheck()
```
minor nit: for consistency with the rest of the file, use `function healthcheck {` instead.

pohly replied:
Fixed.

ixdy commented on hack/local-up-cluster.sh:
```diff
@@ -400,6 +400,44 @@ cleanup()
+    APISERVER_PID=
+  fi
+
+  if [[ -n "${CTRLMGR_PID-}" ]] && ! sudo kill -0 ${CTRLMGR_PID} 2>/dev/null; then
```
this seems to actually be `$CTLRMGR_PID` (and `$CTLRMGR_LOG` below)

pohly replied:
Right, good catch. Fixed.

ixdy commented on hack/local-up-cluster.sh:
```diff
@@ -400,6 +400,44 @@ cleanup()
+  fi
+
+  if [[ -n "$DOCKERIZE_KUBELET" ]]; then
+    # TODO (?): check health also in this case.
```
maybe create an issue and link it here?

pohly replied:
Done, see https://github.com/kubernetes/kubernetes/issues/62474

ixdy commented on hack/local-up-cluster.sh:
```diff
@@ -400,6 +400,44 @@ cleanup()
+
+  if [[ -n "$DOCKERIZE_KUBELET" ]]; then
+    # TODO (?): check health also in this case.
+    :
```
nit - stray `:`?

pohly replied:
No, the : is intentional. A statement is needed between "then" and "fi", and : is the shortest possible statement that doesn't do anything.

ixdy commented on hack/local-up-cluster.sh:
```diff
@@ -400,6 +400,44 @@ cleanup()
+  fi
+
+  if [[ -n "${SCHEDULER_PID-}" ]] && ! sudo kill -0 ${SCHEDULER_PID} 2>/dev/null; then
+    warning "API server terminated unexpectedly, see ${SCHEDULER_LOG}"
```
scheduler, not API server?

pohly replied:
Darn copy-and-paste ;-} You are right of course.

ixdy reviewed (commented):
a few minor nits, otherwise LG

pohly commented:
@ixdy thanks for your review. I think I have addressed your concerns in the rebased and force-pushed commits.

pohly commented:
/test pull-kubernetes-e2e-gce

pohly commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-local-e2e

pohly commented:
/test pull-kubernetes-e2e-kops-aws

pohly commented:
@dims and/or @ixdy: after some retries we currently have a clean test status. Can you help with getting this PR merged?

dims commented:
/assign @cblecker

ixdy commented:
/lgtm
